### ISSUE TITLE
Proposal: combine requires_grad and retain_grad()

### ISSUE THREAD

**Issue by colesbury** [pytorch/pytorch#3625]
Currently, requires_grad means two things: 
1) That we should compute gradients for this variable and functions of this variable
2) On a "leaf" variable, it means we should store the gradient to the "grad" attribute

The `retain_grad()` functions is used to signify that we should store the gradient on non-"leaf" variables to the "grad" attribute.

We should change `requires_grad` so that it signifies that we should store the "grad" attribute on all variables (leaf and non-leaf). We should add a new read-only attribute, `compute_grad`, which serves the first purpose (whether or not we should compute gradients).

 `retain_grad()` can be deprecated and simply set requires_grad=True for backwards compatibility.

**Comment by apaszke** [comment]
Is it really worth changing? The semantics are exactly the same, you just change the name and break a ton of user scripts (`requires_grad` is everywhere)

**Comment by colesbury** [comment]
I think very few user scripts would break: people set requires_grad everywhere, but that remains the same. Only reading requires_grad on non-leaf Variables changes.

**Comment by apaszke** [comment]
Yes, but this checking of `requires_grad` on non-leaf Variables changes. That's what I was thinking about

**Comment by colesbury** [comment]
I've looked through the examples and a few [popular projects](https://github.com/topics/pytorch). I haven't seen any uses like that.

**Comment by saluto** [comment]
I see problems with the current behaviour, too:

```
import torch

# --- example 1:
# If `input.requires_grad = False`, `output.requires_grad_(True)` causes retaining `output.grad`. - Is this expected or not?
# If `input.requires_grad = True`, `output.requires_grad_(True)` does NOT cause retaining `output.grad`. - Is this expected or not?
# At least the two different effects of `output.requires_grad_(True)` based on `input.requires_grad` is confusing and misleading.
input = torch.tensor(5., requires_grad=True)
output = 2 * input
output.requires_grad_(True)
output.backward()
print(input.grad, output.grad)

# --- example 2:
# If `input.requires_grad = True`, `output.retain_grad()` causes retaining `output.grad`. - As expected.
# If `input.requires_grad = False`, `output.retain_grad()` does NOT cause `output.requires_grad = True` (would require additional `output.requires_grad_(True)`) and therefore CAN NOT retain `output.grad`, leading to an exception. - Is this expected or not?
# At least the two different effects of `output.retain_grad()` based on `input.requires_grad` is confusing and misleading.
input = torch.tensor(5., requires_grad=False)
output = 2 * input
output.retain_grad()
output.backward()
print(input.grad, output.grad)
```

I expected, that `output.requires_grad_(True)` and `output.retain_grad()` have an effect on `output.grad` that is independent of `input.requires_grad`. That this is not the case seems really bad to me.

I suggest the following:

1. Remove any ability to change `requires_grad` directly by user (only indirect, see (2.)).
(It should be just a read-only flag, to allow passing the need of `grad_fn` to child tensors, independent of whether the gradient actually should be retained in `.grad`. For retaining, see (2.).)
In tensor factories like `torch.tensor()`, rename the argument `requires_grad` to `retains_grad`, causing both `requires_grad` and `retains_grad` to be set to `True` (same effect as calling `retain_grad_(True)`, see (2.)).
2. Because we removed `requires_grad_(True/False)` (see (1.)), let us define `retain_grad(retain_grad=True)` (and `retain_grad_(retain_grad=True)` as inplace variant) to allow setting `retains_grad = True/False`.
On leaves, this should also set `requires_grad = retains_grad`.
On non-leaves, it should set `requires_grad = retains_grad if not any(input.requires_grad == True for input in self.inputs) else True`. (pseudo code)

I really understand the need for backward compatibility. So let us first make (1.) just a deprecation. I think (2.) is fine. For me that even is the expected behaviour.

For me this seems as quite a fundamental source of confusion and bugs. When to fix this, if not now? Maybe even with version 1.0, considering the introduction of this [post](https://pytorch.org/2018/05/02/road-to-1.0.html) about targeting a stable API? Better early than late.

Thanks for reading. And for helping to develop this incredible piece of my life.

**Comment by gchanan** [comment]
```
# --- example 1:
# If `input.requires_grad = False`, `output.requires_grad_(True)` causes retaining `output.grad`. - Is this expected or not?
# If `input.requires_grad = True`, `output.requires_grad_(True)` does NOT cause retaining `output.grad`. - Is this expected or not?
# At least the two different effects of `output.requires_grad_(True)` based on `input.requires_grad` is confusing and misleading.
```

Yes, these are all expected.  The way to think about the current behavior is `requires_grad` controls whether a gradient is calculated and a grad is only "retained" if retain is set or the variable is a leaf (doesn't have a grad_fn).  The above examples follow from this.

**Comment by saluto** [comment]
The "problem"/bad API I see is that the effect of `requires_grad_()` and `retain_grad()` depends on whether an input `requires_grad` or not. So you cannot rely on either of these calls for non-leaves to just "get" the grad for it on `backward()`. You always have to check whether an input `requires_grad` or you have to call both `requires_grad_()` and `retain_grad()` on the non-leave just to be sure. There is no generic way to signal with one call that "I want the grad for this tensor on every `backward()`".

**Comment by NightMachinery** [comment]
> ```
> # --- example 1:
> # If `input.requires_grad = False`, `output.requires_grad_(True)` causes retaining `output.grad`. - Is this expected or not?
> # If `input.requires_grad = True`, `output.requires_grad_(True)` does NOT cause retaining `output.grad`. - Is this expected or not?
> # At least the two different effects of `output.requires_grad_(True)` based on `input.requires_grad` is confusing and misleading.
> ```
> 
> Yes, these are all expected. The way to think about the current behavior is `requires_grad` controls whether a gradient is calculated and a grad is only "retained" if retain is set or the variable is a leaf (doesn't have a grad_fn). The above examples follow from this.

What happens when a tensor A has `requires_grad` set to False but the gradient is needed for a downstream tensor B that has `requires_grad` set to True? My understanding is that A will have its gradient computed. Will this gradient be saved to `.grad`? I think the gradient will be directly given to `.grad_fn` and not saved, but I am not familiar with the autograd system. What happens if `retain_grad` is True?

What is the use case when someone needs to set `requires_grad` to True but set `retain_grad` to False? I can't see any situations where this would be useful. If it's indeed not useful, then these two are redundant and should become aliases of each other.

**Comment by michaelklachko** [comment]
I remember a use case involving retain_grad attribute, where I wanted to apply regularization to gradients of activations. The idea was to penalize large gradients during training to make the model less sensitive to noise. 

I had to set retain_grad to True here: https://github.com/michaelklachko/NoisyNet/blob/master/noisynet.py#L444 
then compute the grads here: https://github.com/michaelklachko/NoisyNet/blob/master/noisynet.py#L1388
then penalize their growth here: https://github.com/michaelklachko/NoisyNet/blob/master/noisynet.py#L1406

**Comment by NightMachinery** [comment]
> > What is the use case when someone needs to set requires_grad to True but set retain_grad to False?
> 
> This is actually the usual case: every intermediary will have requires_grad to True (their gradient needs to be computed as they are intermediary) but retain_grad to False (as you don't want to save the gradient for it on the .grad field). Note that the main reason for not populating the .grad field as much as possible is to reduce memory usage.

Wouldn't their gradient be computed even if they have `requires_grad` set to False? After all, a common way to freeze parameters is to set their requires_grad to False, while their gradients are still needed for the unfrozen parameters.

**Comment by albanD** [comment]
if requires_grad is False, no gradient will be computed indeed (and retain_grad doesn't really make sense).
